assignment2/task2.py

```diff
@@ -0,0 +1,84 @@
+# -*- coding: UTF-8 -*-
+
+import math
+
+
+class LCABuilder(object):
+    def __init__(self, graph):
+        self.graph = graph
+        self.root = 0
+        self.two_pow = int(math.ceil(math.log(len(graph), 2)))
+
+        self.time_in = [0] * len(graph)
+        self.time_out = [0] * len(graph)
+        self.parent = [[0] * (self.two_pow + 1) for _ in xrange(len(graph))]
+        self.timer = 1
+
+    def build(self):
+        self._dfs(self.root)
+
+    def _dfs(self, ver, pnt=0):
+        self.time_in[ver] = self.timer
+        self.timer += 1
+
+        self.parent[ver][0] = pnt
+        for index in xrange(1, self.two_pow + 1):
+            prev = self.parent[ver][index - 1]
+            self.parent[ver][index] = self.parent[prev][index - 1]
+        for index in xrange(len(self.graph[ver])):
+            to = self.graph[ver][index]
+            if to == pnt:
+                continue
+            self._dfs(to, ver)
+
+        self.time_out[ver] = self.timer
+        self.timer += 1
+
+    def find(self, first_ver, second_ver):
+        if self._check(first_ver, second_ver):
+            return first_ver
+        if self._check(second_ver, first_ver):
+            return second_ver
+        for step in xrange(self.two_pow, -1, -1):
+            if not self._check(self.parent[first_ver][step], second_ver):
+                first_ver = self.parent[first_ver][step]
+        return self.parent[first_ver][0]
+
+    def _check(self, first_ver, second_ver):
+        return (
+            self.time_in[first_ver] <= self.time_in[second_ver] and
+            self.time_out[first_ver] >= self.time_out[second_ver]
+        )
+
```

Comment:
Time complexity: 
* prepocessing: O(n log n)
* get answer for (A, B) request: O(log n)